internal/docstore/dynamodocstore/dynamo.go

```diff
@@ -12,22 +12,134 @@
 	"github.com/aws/aws-sdk-go/service/dynamodb/expression"
+	gcaws "gocloud.dev/aws"
 	"gocloud.dev/gcerrors"
 	"gocloud.dev/internal/docstore"
 	"gocloud.dev/internal/docstore/driver"
 	"gocloud.dev/internal/gcerr"
 )
 
+func init() {
+	docstore.DefaultURLMux().RegisterCollection(Scheme, new(lazySessionOpener))
+}
+
+type lazySessionOpener struct {
+	init   sync.Once
+	opener *URLOpener
+	err    error
+}
+
+func (o *lazySessionOpener) OpenCollectionURL(ctx context.Context, u *url.URL) (*docstore.Collection, error) {
+	o.init.Do(func() {
+		sess, err := session.NewSessionWithOptions(session.Options{SharedConfigState: session.SharedConfigEnable})
+		if err != nil {
+			o.err = err
+			return
+		}
+		o.opener = &URLOpener{
+			ConfigProvider: sess,
+		}
+	})
+	if o.err != nil {
+		return nil, fmt.Errorf("open collection %s: %v", u, o.err)
+	}
+	return o.opener.OpenCollectionURL(ctx, u)
+}
+
+// Scheme is the URL scheme dynamodocstore registers its URLOpener under on
+// docstore.DefaultMux.
+const Scheme = "dynamodocstore"
+
+// URLOpener opens dynamodocstore URLs like
+// "dynamodocstore://mytable?partition_key=partkey&sort_key=sortkey".
+//
+// The URL Host is used as the table name. See
+// https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/WorkingWithTables.html
+// for more details.
+//
+// The following query parameters are supported:
+//
+//   - partition_key (required): the path to the partition key of a table or an index.
+//   - sort_key: the path to the sort key of a table or an index.
+//
+// See https://godoc.org/gocloud.dev/aws#ConfigFromURLParams for supported query
+// parameters for overriding the aws.Session from the URL.
+type URLOpener struct {
+	// ConfigProvider must be set to a non-nil value.
+	ConfigProvider client.ConfigProvider
+}
+
+// OpenCollectionURL opens the collection at the URL's path. See the package doc for more details.
+func (o *URLOpener) OpenCollectionURL(ctx context.Context, u *url.URL) (*docstore.Collection, error) {
+	q := u.Query()
+
+	partitionKey := q.Get("partition_key")
+	if partitionKey == "" {
+		return nil, fmt.Errorf("open collection %s: partition_key is required to open a table", u)
+	}
+	q.Del("partition_key")
+	sortKey := q.Get("sort_key")
+	q.Del("sort_key")
+
+	tableName := u.Host
+	if tableName == "" {
+		return nil, fmt.Errorf("open collection %s: URL's host cannot be empty (the table name)", u)
+	}
+	if u.Path != "" {
+		return nil, fmt.Errorf("open collection %s: URL path must be empty, only the host is needed", u)
+	}
+
+	configProvider := &gcaws.ConfigOverrider{
+		Base: o.ConfigProvider,
+	}
+	overrideCfg, err := gcaws.ConfigFromURLParams(q)
+	if err != nil {
+		return nil, fmt.Errorf("open collection %s: %v", u, err)
+	}
+	configProvider.Configs = append(configProvider.Configs, overrideCfg)
+	db, err := Dial(configProvider)
+	if err != nil {
+		return nil, fmt.Errorf("open collection %s: %v", u, err)
+	}
+	return OpenCollection(db, tableName, partitionKey, sortKey), nil
```

Comment:
OK for separate PR, but `OpenCollection` needs an `Options`, which should then be added here and to `URLOpener`.

Reply:
Ack, filed #1877